source/common/access_log/access_log_formatter.cc

```diff
@@ -661,12 +935,39 @@ std::string FilterStateFormatter::format(const Http::HeaderMap&, const Http::Hea
     // TODO(lizan): add support of unknown Any.
     return UnspecifiedValueString;
   }
-  if (max_length_.has_value() && value.length() > max_length_.value()) {
-    return value.substr(0, max_length_.value());
-  }
+
+  truncate(value, max_length_);
   return value;
 }
 
+ProtobufWkt::Value
+FilterStateFormatter::formatValue(const Http::HeaderMap&, const Http::HeaderMap&,
+                                  const Http::HeaderMap&,
+                                  const StreamInfo::StreamInfo& stream_info) const {
+  ProtobufTypes::MessagePtr proto = filterState(stream_info);
+  if (proto == nullptr) {
+    return unspecifiedValue();
+  }
+
+  std::string json;
+  const auto status = Protobuf::util::MessageToJsonString(*proto, &json);
+  if (!status.ok()) {
+    // If the message contains an unknown Any (from WASM or Lua), MessageToJsonString will fail.
+    // TODO(lizan): add support of unknown Any.
+    return unspecifiedValue();
+  }
+
+  // TODO(zuercher): find a way to perform this without converting to JSON and back.
```

Comment:
use `MessageUtil::jsonConvert`

Reply:
I can expand the comment to be clearer, but the reason jsonConvert doesn't work is that the output is not necessary a ProtobufWkt::Struct, it can be a ProtobufWkt::Value.

Reply:
hmm, I'm sure the `jsonConvertInternal` does work, so probably just add another variant there to take `ProtobufWkt::Value` as dest?

Reply:
Done.